media/multipart/src/main/java/io/helidon/media/multipart/MultiPartDecoder.java

```diff
@@ -106,120 +142,231 @@ public void onNext(DataChunk chunk) {
+        }
     }
 
     @Override
     public void onComplete() {
-        initFuture.whenComplete((e, t) -> {
-            if (upstream != SubscriptionHelper.CANCELED) {
-                upstream = SubscriptionHelper.CANCELED;
-                try {
-                    parser.close();
-                } catch (MimeParser.ParsingException ex) {
-                    emitter.fail(ex);
-                    releaseChunks();
-                }
-            }
-        });
+        if (upstream != SubscriptionHelper.CANCELED) {
+            upstream = SubscriptionHelper.CANCELED;
+            drain();
+        }
+    }
+
+    private boolean halfInit(int mask) {
+        // Attempts to set the given init mask, if contenders is in the right state for that, and
+        // reports whether the contenders was in a state where that part of init needed completing.
+        int c = contenders.getAndUpdate(v -> v < 0 ? v | mask : v);
+        return c < 0 && (c & mask) == 0;
     }
 
     private void deferredInit() {
-        if (upstream != null && downstream != null) {
-            emitter = BufferedEmittingPublisher.create();
-            emitter.onRequest(this::onPartRequest);
-            emitter.onEmit(this::drainPart);
-            //emitter.onCancel(this::onPartCancel);
-            emitter.subscribe(downstream);
-            initFuture.complete(emitter);
-            downstream = null;
+        // deferredInit is invoked twice: onSubscribe and subscribe
+        // after onSubscribe and subscribe three top bits are set
+        // adding SUBSCRIPTION_LOCK for the first time sets the fourth bit
+        // adding SUBSCRIPTION_LOCK for the second time clears all top bits
+        // making the contenders counter 0,
+        // unless there were onError, onComplete, request for parts or downstream cancellation
+        if (contenders.addAndGet(SUBSCRIPTION_LOCK) > 0) {
+            drainLoop();
         }
     }
 
-    private void onPartRequest(long requested, long total) {
-        // require more raw chunks to decode if the decoding has not
-        // yet started or if more data is required to make progress
-        if (!parserEventProcessor.isStarted() || parserEventProcessor.isDataRequired()) {
-            upstream.request(1);
-        }
+    private long partsRequested() {
+        // Returns a negative number, if we are serving outer Subscriber, and we can tell it may
+        // not issue more requests for parts: either cancelled, or a bad request was observed.
+        // Otherwise returns a positive number, if serving inner Subscriber, or actual partsRequested.
+        return bodyPartPublisher != null ? 1 : partsRequested.get();
     }
 
-    private void onPartCancel() {
-        emitter.clearBuffer(this::drainPart);
+    private void cleanup() {
+        // drop the reference to parserIterator, but keep it safe for any later invocation of parserIterator
+        parserIterator = EMPTY_PARSER_ITERATOR;
+        error = null;
```

Comment:
I apologise I didn't make the comment 
> Add upstream = CANCELED; to cleanup

clearer and forgot to add it in the `onError` synchronization fixup.

Please, add `upstream = CANCELED` somewhere here to drop references to upstream.